ccl/multitenantccl/tenantcostclient: TestConsumption failed

ccl/multitenantccl/tenantcostclient.TestConsumption [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8095788?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8095788?buildTab=artifacts#/) on release-22.2 @ [ea0e966fe05a788fd123a08cec8ef5e935fa5496](https://github.com/cockroachdb/cockroach/commits/ea0e966fe05a788fd123a08cec8ef5e935fa5496):


```
=== RUN   TestConsumption
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/0cce69e026224ea71be99da3223b988f/logTestConsumption57809745
    test_log_scope.go:79: use -show-logs to present logs inline
    tenant_side_test.go:885: usage after write: r_u:35.13038335253964 kv_r_u:5.1303863525390625 read_batches:1 read_requests:34 read_bytes:24929 sql_pods_cpu_seconds:0.09000000000000075 
    tenant_side_test.go:909: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/0cce69e026224ea71be99da3223b988f/logTestConsumption57809745
--- FAIL: TestConsumption (2.23s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #92450 ccl/multitenantccl/tenantcostclient: TestConsumption failed [C-test-failure O-robot branch-release-22.1]
</p>
</details>
/cc @cockroachdb/multi-tenant
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConsumption.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-22797

ccl/multitenantccl/tenantcostclient.TestConsumption [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8148502?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8148502?buildTab=artifacts#/) on release-22.2 @ [07a53a36601e9ca5fcffcff55f69b43c6dfbf1c1](https://github.com/cockroachdb/cockroach/commits/07a53a36601e9ca5fcffcff55f69b43c6dfbf1c1):


```
=== RUN   TestConsumption
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/0cce69e026224ea71be99da3223b988f/logTestConsumption2414672614
    test_log_scope.go:79: use -show-logs to present logs inline
    tenant_side_test.go:885: usage after write: r_u:13.333332000000155 sql_pods_cpu_seconds:0.040000000000000036 
    tenant_side_test.go:909: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/0cce69e026224ea71be99da3223b988f/logTestConsumption2414672614
--- FAIL: TestConsumption (1.74s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #92450 ccl/multitenantccl/tenantcostclient: TestConsumption failed [C-test-failure O-robot branch-release-22.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConsumption.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/multitenantccl/tenantcostclient.TestConsumption [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10282761?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10282761?buildTab=artifacts#/) on release-22.2 @ [afe2d46600257f6c022ce89168bebdc0023c8215](https://github.com/cockroachdb/cockroach/commits/afe2d46600257f6c022ce89168bebdc0023c8215):


```
=== RUN   TestConsumption
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/0cce69e026224ea71be99da3223b988f/logTestConsumption1792693212
    test_log_scope.go:79: use -show-logs to present logs inline
    tenant_side_test.go:891: usage after write: r_u:289.5164809980471 kv_r_u:202.84982299804688 read_batches:6 read_requests:3 read_bytes:14 write_batches:1 write_requests:18 write_bytes:184806 sql_pods_cpu_seconds:0.2600000000000007 
    tenant_side_test.go:915: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/0cce69e026224ea71be99da3223b988f/logTestConsumption1792693212
--- FAIL: TestConsumption (2.79s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #92450 ccl/multitenantccl/tenantcostclient: TestConsumption failed [C-test-failure O-robot branch-release-22.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConsumption.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go
+++ b/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go
@@ -873,8 +873,9 @@ func TestConsumption(t *testing.T) {
 	hostServer, _, _ := serverutils.StartServer(t, base.TestServerArgs{DefaultTestTenant: base.TestControlsTenantsExplicitly})
 	defer hostServer.Stopper().Stop(context.Background())
 
+	const targetPeriod = time.Millisecond
 	st := cluster.MakeTestingClusterSettings()
-	tenantcostclient.TargetPeriodSetting.Override(context.Background(), &st.SV, time.Millisecond)
+	tenantcostclient.TargetPeriodSetting.Override(context.Background(), &st.SV, targetPeriod)
 	tenantcostclient.CPUUsageAllowance.Override(context.Background(), &st.SV, 0)
 
 	testProvider := newTestProvider()
@@ -902,21 +903,32 @@ func TestConsumption(t *testing.T) {
 		r.Exec(t, "INSERT INTO t (v) SELECT repeat('1234567890', 1024) FROM generate_series(1, 10) AS g(i)")
 		const expectedBytes = 10 * 10 * 1024
 
-		afterWrite := testProvider.waitForConsumption(t)
-		delta := afterWrite
-		delta.Sub(&beforeWrite)
-		if delta.WriteBatches < 1 || delta.WriteRequests < 2 || delta.WriteBytes < expectedBytes*2 {
-			t.Errorf("usage after write: %s", delta.String())
-		}
+		// Try a few times because background activity can trigger bucket
+		// requests before the test query does.
+		testutils.SucceedsSoon(t, func() error {
+			afterWrite := testProvider.waitForConsumption(t)
+			delta := afterWrite
+			delta.Sub(&beforeWrite)
+			if delta.WriteBatches < 1 || delta.WriteRequests < 2 || delta.WriteBytes < expectedBytes*2 {
+				return errors.Newf("usage after write: %s", delta.String())
+			}
+			return nil
+		})
 
+		beforeRead := testProvider.waitForConsumption(t)
 		r.QueryStr(t, "SELECT min(v) FROM t")
 
-		afterRead := testProvider.waitForConsumption(t)
-		delta = afterRead
-		delta.Sub(&afterWrite)
-		if delta.ReadBatches < 1 || delta.ReadRequests < 1 || delta.ReadBytes < expectedBytes {
-			t.Errorf("usage after read: %s", delta.String())
-		}
+		// Try a few times because background activity can trigger bucket
+		// requests before the test query does.
+		testutils.SucceedsSoon(t, func() error {
+			afterRead := testProvider.waitForConsumption(t)
+			delta := afterRead
+			delta.Sub(&beforeRead)
+			if delta.ReadBatches < 1 || delta.ReadRequests < 1 || delta.ReadBytes < expectedBytes {
+				return errors.Newf("usage after read: %s", delta.String())
+			}
+			return nil
+		})
 		r.Exec(t, "DELETE FROM t WHERE true")
 	}
 	// Make sure some CPU usage is reported.

